### PR TITLE
Add ability to build and push Docker image during interactive `prefect deploy`

### LINKED ISSUE #9910: Prompt for Docker information when a Docker-based infrastructure type is selected during interactive `prefect deploy` run
### Summary

When a user has selected a Docker-based infrastructure type (e.g. `docker`, `ecs`, `cloud-run`) during an interactive `prefect deploy` session, prompt the user for information about building and pushing a Docker image for their deployed flow.

### Details

The CLI can determine if a work pool supports Docker images by inspecting its `base_job_template`. If the work pool has a `image` variable, it can be assumed to support use of custom Docker images.

If the user selects a Docker-based infrastructure type, the CLI should follow the following flow:
1. Ask the user if they would like to build a custom Docker image.
2. If a Dockerfile is present in the current directory, ask them if they would like to use that Dockerfile
3. Ask for an image name
4. Ask for an image tag
5. Ask the user if they would like to push this image to a remote registry
6. Ask which registry they would like to push to
7. Ask if it is a private registry
8. Ask for credentials to use for pushing

The answers to these questions should be use to configure the necessary steps to build and push a Docker image. If a Dockerfile is absent, the auto-build feature available in the `build_docker_image` should be used. A `DockerRegistryCredentials` block should be created if credentials are provided.

### PR DESCRIPTION
**Pull request by serinamarie**
Closes #9910

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.

### CODE REVIEW

File: src/prefect/cli/deploy.py [hunk 1]
```diff
@@ -428,6 +430,27 @@ async def _run_single_deploy(
             console=app.console, client=client
         )
 
+    if is_interactive() and not deploy_config.get("build"):
```
Comment (serinamarie) [inline src/prefect/cli/deploy.py:433]: should be specifically checking for `build_docker_image` step

File: src/prefect/cli/deploy.py [hunk 2]
```diff
@@ -428,6 +430,27 @@ async def _run_single_deploy(
...
+        # currently, will only hit if you use an existing deployment configuration
+        # a new one won't have the deploy_config with an image, just a default
+        if get_from_dict(deploy_config, "work_pool.job_variables.image"):
```
Comment (serinamarie) [inline src/prefect/cli/deploy.py:436]: access directly from work_pool base job template

File: src/prefect/cli/_prompts.py [hunk 3]
```diff
@@ -31,6 +34,10 @@
 
 from prefect.client.schemas.actions import WorkPoolCreate
 
+from prefect.utilities.slugify import slugify
+
+from prefect_docker import DockerRegistryCredentials
```
Comment (serinamarie) [inline src/prefect/cli/_prompts.py:39]: use `DockerRegistry` instead

**Review by desertaxle — COMMENTED** [review]
This is looking great so far! I got a Docker image built and pushed after a few tweaks. I added comments for the tweaks that I had to make.

One thing that is still needed is setting the correct pull step for the Docker container. If we do an auto-build, we can set the pull step the same way the `docker` recipes do, but it's a little trickier if the user uses a custom Docker image without auto-build. I think we either need to ask for the path we should change to when running the deployment, or direct them towards pulling their code from a remote location.

File: src/prefect/cli/_prompts.py [hunk 4]
```diff
@@ -344,6 +350,119 @@ async def prompt_select_work_pool(
...
+        return []
+
+    push_step = {
+        "requires": "prefect-docker",
+        "image_name": "{{ build-image.image_name }}",
+        "tag": "{{ build-id.tag }}",
+    }
+
+    docker_credentials = {}
+    docker_credentials["registry_url"] = prompt("Registry URL", default="default")
```
Comment (serinamarie) [inline src/prefect/cli/_prompts.py:419]:
```suggestion
    docker_credentials["registry_url"] = prompt("Registry URL")
```

File: src/prefect/cli/deploy.py [hunk 5]
```diff
@@ -428,6 +430,34 @@ async def _run_single_deploy(
...
+            )
+            if build_docker_image_step != []:
+                deploy_config["build"] = build_docker_image_step
+                push_docker_image_step = await prompt_push_custom_docker_image(
+                    app.console, deploy_config
+                )
+                if push_docker_image_step != []:
+                    deploy_config["build"][0][
+                        "prefect_docker.deployments.steps.build_docker_image"
+                    ]["push"] = True
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/deploy.py]: Also makes sense 👍

File: src/prefect/cli/_prompts.py [hunk 6]
```diff
@@ -344,6 +351,121 @@ async def prompt_select_work_pool(
...
+        return []
+
+    push_step = {
+        "requires": "prefect-docker",
+        "image_name": "{{ build-id.image_name }}",
+        "tag": "{{ build-id.tag }}",
+    }
+
+    docker_credentials = {}
+    docker_credentials["registry_url"] = prompt("Registry URL", default="default")
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/_prompts.py]: Hm, yeah, I removed the default URL :)

File: src/prefect/cli/_prompts.py [hunk 7]
```diff
@@ -344,6 +351,121 @@ async def prompt_select_work_pool(
...
+                # this will otherwise raise when build steps are run as the auto-build feature
+                # executed in the build_docker_image step will create a temporary Dockerfile
+                raise ValueError(
+                    "A Dockerfile already exists. Please remove or rename the existing"
+                    " one."
+                )
+    else:
+        build_step["dockerfile"] = "auto"
+
+    build_step["image_name"] = prompt("Image name", default="default")
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/_prompts.py]: Good ideas! I've applied your suggestions.

File: src/prefect/cli/_prompts.py [hunk 8]
```diff
@@ -344,6 +351,121 @@ async def prompt_select_work_pool(
...
+                    default="default",
+                    password=True,
+                )
+            except GetPassWarning:
+                docker_credentials["password"] = prompt(
+                    "Docker registry password",
+                    console=console,
+                    default="default",
+                )
+            await DockerRegistryCredentials(
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/_prompts.py]: Good idea! I have given the user the option to install `prefect-docker` if they'd like to use `DockerRegistryCredentials`.

**Comment by desertaxle** [comment]
One thing that I missed on my first review, we want to make sure we're setting an `image` variable in the `job_variables` section under `work_pool` when there is a `build_docker_image` step in the `build` deploy action.

If a user runs `prefect deploy` and chooses to build a custom Docker image, their generated deploy config should look like this:
```yaml
work_pool:
    name: ...
    job_variables:
        image: "{{ build-image.image }}"
```

To determine if we need to set `image` in `job_variables`, we can check if the following three statements are true:
- The `build` deploy action has a `build_docker_image` step
- The selected/created work pool has an `image` variable in its base job template
- There is not an existing `image` key in `job_variables`

If all three are true, we should set the `image` key in `job_variables`.

**Comment by serinamarie** [comment]
> One thing that is still needed is setting the correct pull step for the Docker container. If we do an auto-build, we can set the pull step the same way the docker recipes do, but it's a little trickier if the user uses a custom Docker image without auto-build. I think we either need to ask for the path we should change to when running the deployment, or direct them towards pulling their code from a remote location.

Thanks! This is addressed in: https://github.com/PrefectHQ/prefect/pull/10090

File: src/prefect/cli/deploy.py [hunk 9]
```diff
@@ -433,6 +435,51 @@ async def _run_single_deploy(
...
+                else:
+                    actions["build"] = [build_docker_image_step]
+
+                work_pool_job_variables_image_not_found = not get_from_dict(
+                    deploy_config, "work_pool.job_variables.image"
+                )
+                if work_pool_job_variables_image_not_found:
+                    deploy_config["work_pool"]["job_variables"][
+                        "image"
+                    ] = "{{ build-image.image }}"
```
Comment (serinamarie) [inline src/prefect/cli/deploy.py:469]: If not found, I should also update the `work_pool` itself so it reflects in the UI.

**Comment by serinamarie** [comment]
> One thing that I missed on my first review, we want to make sure we're setting an `image` variable in the `job_variables` section under `work_pool` when there is a `build_docker_image` step in the `build` deploy action.
> 
> If a user runs `prefect deploy` and chooses to build a custom Docker image, their generated deploy config should look like this:
> 
> ```yaml
> work_pool:
>     name: ...
>     job_variables:
>         image: "{{ build-image.image }}"
> ```
> 
> To determine if we need to set `image` in `job_variables`, we can check if the following three statements are true:
> 
> * The `build` deploy action has a `build_docker_image` step
> * The selected/created work pool has an `image` variable in its base job template
> * There is not an existing `image` key in `job_variables`
> 
> If all three are true, we should set the `image` key in `job_variables`.

Thanks! That should be addressed here, although it may not be the ideal way. It updates the `prefect.yaml` under the deployments section for that particular deployment. I will also want to add tests. 
[67f8724](https://github.com/PrefectHQ/prefect/pull/10022/commits/67f872418816dd3ac7a394b5ceb4607706d09013)

**Comment by serinamarie** [comment]
These tests do import `prefect-docker`, which would be best to avoid if possible:
```
test_prompt_push_docker_image_accepted_private_registry_use_existing_prefect_docker_creds
test_prompt_push_docker_image_accepted_private_registry_use_new_prefect_docker_creds
test_prompt_push_docker_image_accepted_private_registry_reject_use_existing_prefect_docker_creds
```

Options:
1. We could leave these tests out for now
2. Mock the `prefect_docker.DockerRegistryCredentials` block

**Review by desertaxle — COMMENTED** [review]
I got the full flow working with a couple of tweaks, and it was awesome! I think we're just a few changes away from this being ready to merge!

File: src/prefect/cli/deploy.py [hunk 10]
```diff
@@ -466,6 +513,16 @@ async def _run_single_deploy(
...
 
+    if update_work_pool_image:
+        if "build-image" not in step_outputs:
+            app.console.print(
+                "Warning: no build-image step found in the deployment build steps."
+                " The work pool image will not be updated."
+            )
+        deploy_config["work_pool"]["job_variables"]["image"] = step_outputs[
+            "build-image"
+        ]["image"]
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/deploy.py:524]: Yeah, that makes sense!

File: src/prefect/cli/_prompts.py [hunk 11]
```diff
@@ -345,6 +353,150 @@ async def prompt_select_work_pool(
...
+        return
+
+    push_step = {
+        "requires": "prefect-docker>=0.3.1",
+        "image_name": "{{ build-image.image_name }}",
+        "tag": "{{ build-image.tag }}",
+    }
+
+    registry_url = prompt("Registry URL")
+    push_step["registry_url"] = registry_url
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/_prompts.py]: Got it. It should now be
```python
build_step["image_name"] = f"{registry_url}/{build_step['image_name']}"
```

Hopefully that's what you meant.

File: src/prefect/cli/deploy.py [hunk 12]
```diff
@@ -466,6 +513,16 @@ async def _run_single_deploy(
 
     step_outputs.update(variable_overrides)
 
+    if update_work_pool_image:
+        if "build-image" not in step_outputs:
+            app.console.print(
+                "Warning: no build-image step found in the deployment build steps."
+                " The work pool image will not be updated."
+            )
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/deploy.py:524]: 👍

[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/deploy.py:524]: Probably won't make into this PR.

**Comment by serinamarie** [comment]
>  got the full flow working with a couple of tweaks, and it was awesome! I think we're just a few changes away from this being ready to merge!

That's awesome, glad to hear that! Let me know if there's any other changes you think I should make.

**Review by desertaxle — APPROVED** [review]
🚀 🚀 🚀

**Pull request merged**